src/psyclone/psyir/transformations/scalarization_trans.py

```diff
@@ -0,0 +1,324 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. B. G. Chalk, STFC Daresbury Lab
+
+'''This module provides the sclarization transformation class.'''
+
+import itertools
+
+from psyclone.core import VariablesAccessInfo
+from psyclone.psyGen import Kern
+from psyclone.psyir.nodes import Assignment, Call, CodeBlock, IfBlock, \
+        Loop, Node, Reference, Routine, WhileLoop
+from psyclone.psyir.symbols import DataSymbol
+from psyclone.psyir.transformations.loop_trans import LoopTrans
+
+
+class ScalarizationTrans(LoopTrans):
+    '''This transformation takes a Loop and converts any array accesses
+    to scalar if the results of the loop are unused, and the initial value
+    is unused. For example:
+
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.transformations import ScalarizationTrans
+    >>> from psyclone.psyir.nodes import Loop
+    >>> code = """program test
+    ... integer :: i,j
+    ... real :: a(100), b(100)
+    ... do i = 1,100
+    ...   a(i) = i
+    ...   b(i) = a(i) * a(i)
+    ... end do
+    ... do j = 1, 100
+    ...  if(b(i) > 200) then
+    ...    print *, b(i)
+    ...  end if
+    ... end do
+    ... end program"""
+    >>> psyir = FortranReader().psyir_from_source(code)
+    >>> scalarise = ScalarizationTrans()
+    >>> scalarise.apply(psyir.walk(Loop)[0])
+    >>> print(FortranWriter()(psyir))
+    program test
+      integer :: i
+      integer :: j
+      real, dimension(100) :: a
+      real, dimension(100) :: b
+      real :: a_scalar
+      <BLANKLINE>
+      do i = 1, 100, 1
+        a_scalar = i
+        b(i) = a_scalar * a_scalar
+      enddo
+      do j = 1, 100, 1
+        if (b(i) > 200) then
+          ! PSyclone CodeBlock (unsupported code) reason:
+          !  - Unsupported statement: Print_Stmt
+          PRINT *, b(i)
+        end if
+      enddo
+    <BLANKLINE>
+    end program test
+    '''
+
+    @staticmethod
+    def _is_local_array(signature, var_accesses):
```

Comment:
Just to let you know that now we allow type hints e.g. 
`def _is_local_array(signature: Signature, var_accesses: VariablesAccessInfo) -> bool:`

instead of the ":type:" docstrings. Although not mandatory for now it is more readable.

Reply:
I've added them alongside the type docstrings - I can remove them if you want but I wasn't sure if we were replacing or using alongside.

Reply:
Sorry for not clarifying it, we are replacing them, so you can delete the docstrings types unless is something currently not representable as typehint

Reply:
Docstring types removed where appropriate.